python/pyarrow/table.pxi

```diff
@@ -1450,8 +1450,129 @@ cdef _sanitize_arrays(arrays, names, schema, metadata,
             converted_arrays.append(item)
     return converted_arrays
 
+cdef class _Tabular(_PandasConvertible):
+    """Internal: An interface for common operations on tabular objects."""
 
-cdef class RecordBatch(_PandasConvertible):
+    def __init__(self):
+        raise TypeError("This object is not instantiable, "
+                        "use a subclass instead.")
+
+    def __repr__(self):
+        if not self._is_initialized():
+            raise ValueError("This object's internal pointer is NULL, do not "
```

Comment:
Looking back at when this was added:
- https://github.com/apache/arrow/pull/6067
- https://github.com/apache/arrow/pull/6067/files#diff-cede36e8e2e0eb6e6e1ee21745db9687174527f463520c6e6d8b9e8f957bf304

it might make sense to keep it but note I am not familiar with the self-destruct option for table.

Reply:
Yes, you can have a table object that no longer is being backed by Arrow data when doing `df = table.to_pandas(self_destruct=True)`. This check and error then prevents getting a segfault when just printing `table` (calling any other method on it will still segfault)

This options seems to have no effect for RecordBatch (I assume this is because the RecordBatch.to_pandas method converts the batch into a Table and then calls Table.to_pandas, so even though the Table C++ object is destructed, the original RecordBatch still owns that data)

Reply:
So we need to keep this check, and I think it is fine that for RecordBatch is this essentially never used